systemtest/infra_kafka.go

```diff
@@ -0,0 +1,230 @@
+		}
+	}
+	logger().Infof("provisioning Kafka infrastructure with config: %+v", cfg)
+	tf, err := NewTerraform(ctx, cfg.TFPath)
+	if err != nil {
+		return err
+	}
+	if err := tf.Init(ctx, tfexec.Upgrade(true)); err != nil {
+		return err
+	}
+	// Create a dummy topic so the infrastructure awaits until all the required
+	// infrastructure is available. It will be destroyed during the first test.
+	if len(cfg.Topics) == 0 {
+		cfg.Topics = append(cfg.Topics, SuffixTopics("dummy")...)
+	}
+	jsonTopics, err := json.Marshal(cfg.Topics)
+	if err != nil {
+		return err
+	}
+	namespaceVar := tfexec.Var(fmt.Sprintf("namespace=%s", cfg.Namespace))
+	nameVar := tfexec.Var(fmt.Sprintf("name=%s", cfg.Name))
+	topicsVar := tfexec.Var(fmt.Sprintf("topics=%s", jsonTopics))
+	// Ensure terraform destroy runs once per TF path.
+	RegisterDestroy(cfg.TFPath, func() {
+		logger().Info("destroying provisioned Kafka infrastructure...")
+		tf.Destroy(ctx, topicsVar, namespaceVar, nameVar)
+	})
+	if err := tf.Apply(ctx, topicsVar, namespaceVar, nameVar); err != nil {
+		return err
+	}
+	tfOutput, err := tf.Output(ctx)
+	if err != nil {
+		return err
+	}
+	if raw, ok := tfOutput[KafkaBrokersKey]; ok {
+		var brokers []string
+		if err := json.Unmarshal(raw.Value, &brokers); err != nil {
+			return err
+		}
+		if len(brokers) > 0 {
+			SetKafkaBrokers(brokers...)
+		}
+	}
+	dt := string(bytes.TrimSpace(tfOutput[KafkaDeploymentTypeKey].Value))
+	// If PortForward is set to true, and the deployment_type is k8s, start
+	// forwarding traffic to/from kafka to localhost.
+	if cfg.PortForward && dt == `"k8s"` {
+		var err error
+		runOnce.Do(func() {
+			// NOTE(marclop) These assume terraform uses the strimzi operator.
+			if cfg.PortForwardResource == "" {
+				cfg.PortForwardResource = fmt.Sprintf("service/%s-kafka-bootstrap", cfg.Name)
+			}
+			if cfg.PortForwardMapping == "" {
+				cfg.PortForwardMapping = "9093:9093"
+			}
+			err = kubectlForwardKafka(cfg.Namespace,
+				cfg.Name,
+				cfg.PortForwardResource,
+				cfg.PortForwardMapping,
+			)
+		})
+		if err != nil {
+			return err
+		}
+	}
+	logger().Info("Kafka infastructure fully provisioned!")
+	return nil
+}
+
+func kubectlForwardKafka(ns, name, service, mapping string) error {
+	cmd := exec.Command("kubectl", "port-forward", "-n", ns, service, mapping)
+	logger().Infof("Running %s...", strings.Join(cmd.Args, " "))
+	if err := cmd.Start(); err != nil {
+		return err
+	}
+	// Patch zone for Kafka client calls to be correctly resolved.
+	srv, err := mockdns.NewServerWithLogger(map[string]mockdns.Zone{
+		// NOTE(marclop) Assumes terraform uses the strimzi operator.
+		fmt.Sprintf("%s-kafka-0.%s-kafka-brokers.%s.svc.", name, name, ns): {
+			A: []string{"127.0.0.1"},
+		},
+	}, log.New(io.Discard, "", 0), false)
+	if err != nil {
+		return err
+	}
+	if mockResolver == nil {
+		resolver := new(net.Resolver)
+		srv.PatchNet(resolver)
+		mockResolver = resolver
+	}
+	go cmd.Wait() // Run the forwarder
```

Comment:
I am bit concerned about this, port-forwarder is not that reliable as it doesn't reconnect/retry. If this was to fail in between the test then we would not know about it and failures would not be that informative.